torch/hub.py

```diff
@@ -119,29 +121,82 @@ def _read_url(url):
+    # https://github.com/NicolasHug/vision/commit/8949c7011facf7801fdf077cc3e4ecd8f0940c7e
+    # So we want avoid downloading code that can come from a potentionally malicious fork.
+
+    try:
+        # The above issue only exists for commit hashes. So if we know that ``branch``
+        # isn't a commit hash, we can return early. This should avoid many GitHub API calls.
+        # If we can't convert branch to a hex int we know for sure it's not a commit hash
+        int(branch, 16)
+    except ValueError:
+        return
+    # Note: here it's still possible that the branch param corresponds to a branch name
+    # or a tag name, so we need to check for those as well.
+
+    def branch_belongs_to_repo():
+        # Return True if either:
+        # - branch corresponds to a branch name in the repo
+        # - branch corresponds to a tag name in the repo
+        # - branch corresponds to a commit that has an associated tag in the repo. Technically,
+        #   this is an undocumented feature: we explicitly tell users that they should not pass
+        #   commit hashes.
+
+        def find_in_refs(ref_kind):
+            # We limit the search to 5k branches / tags, which should be more than enough
+            for page in range(1, 50):
+                url = f'https://api.github.com/repos/{repo_owner}/{repo_name}/{ref_kind}?per_page=100&page={page}'
+                response = json.loads(_read_url(Request(url, headers=headers)))
+                # Empty response means no more data to process
+                if not response:
+                    return False
+                for br in response:
+                    if br['name'] == branch or br['commit']['sha'].startswith(branch):
+                        return True
+
+            # Note: this should never be executed, unless a repo really has 5k+ branches/tags
+            warnings.warn(
+                f"Torchhub tried to validate {branch} and looked at 5000 {ref_kind}, "
+                "but couldn't find a match."
+            )
+            return False
+
+        headers = {'Accept': 'application/vnd.github.v3+json'}
+        token = os.environ.get(ENV_GITHUB_TOKEN)
+        if token is not None:
+            headers['Authorization'] = f'token {token}'
+        return any(find_in_refs(ref_kind) for ref_kind in ('branches', 'tags'))
+
+    try:
+        branch_found = branch_belongs_to_repo()
+    except HTTPError as e:
+        if e.code == 403 and "rate limit exceeded" in str(e):
+            raise ValueError(
+                f"Torchhub was unable to verify that {branch} is indeed part of the {repo_owner}/{repo_name} repo, "
+                f"as it received a GitHub API rate limit error. Consider setting the {ENV_GITHUB_TOKEN} env variable "
+                "to a GitHub token that has more API credits. "
+                "If you're absolutely sure about what you are doing, you can bypass this security check by setting "
+                "skip_validation=True."
+            )
```

Comment:
maybe you want to reset the exception context.

```suggestion
            ) from None
```

Reply:
Good point, actually I think we should do `from e` to properly preserve the whole stack (it provides more details for the user while still keeping the stack clean)